include/SFML/Graphics/Color.inl

```diff
@@ -0,0 +1,149 @@
+//    and must not be misrepresented as being the original software.
+//
+// 3. This notice may not be removed or altered from any source distribution.
+//
+////////////////////////////////////////////////////////////
+
+
+////////////////////////////////////////////////////////////
+constexpr inline Color::Color() :
+r(0),
+g(0),
+b(0),
+a(255)
+{
+
+}
+
+
+////////////////////////////////////////////////////////////
+constexpr inline Color::Color(Uint8 red, Uint8 green, Uint8 blue, Uint8 alpha) :
+r(red),
+g(green),
+b(blue),
+a(alpha)
+{
+
+}
+
+
+////////////////////////////////////////////////////////////
+constexpr inline Color::Color(Uint32 color) :
+r(static_cast<Uint8>((color & 0xff000000) >> 24)),
+g(static_cast<Uint8>((color & 0x00ff0000) >> 16)),
+b((color & 0x0000ff00) >> 8 ),
+a((color & 0x000000ff) >> 0 )
+{
+
+}
+
+
+////////////////////////////////////////////////////////////
+constexpr inline Uint32 Color::toInteger() const
+{
+    return static_cast<Uint32>((r << 24) | (g << 16) | (b << 8) | a);
+}
+
+
+////////////////////////////////////////////////////////////
+constexpr inline bool operator ==(const Color& left, const Color& right)
+{
+    return (left.r == right.r) &&
+           (left.g == right.g) &&
+           (left.b == right.b) &&
+           (left.a == right.a);
+}
+
+
+////////////////////////////////////////////////////////////
+constexpr inline bool operator !=(const Color& left, const Color& right)
+{
+    return !(left == right);
+}
+
+
+////////////////////////////////////////////////////////////
+constexpr inline Color operator +(const Color& left, const Color& right)
+{
+    const auto clampedAdd = [](Uint8 lhs, Uint8 rhs) -> Uint8
+    {
+        const int intResult = static_cast<int>(lhs) + static_cast<int>(rhs);
+        return static_cast<Uint8>(intResult < 255 ? intResult : 255);
+    };
+
+    return Color(clampedAdd(left.r, right.r),
+                 clampedAdd(left.g, right.g),
+                 clampedAdd(left.b, right.b),
+                 clampedAdd(left.a, right.a));
+}
+
+
+////////////////////////////////////////////////////////////
+constexpr inline Color operator -(const Color& left, const Color& right)
+{
+    const auto clampedSub = [](Uint8 lhs, Uint8 rhs) -> Uint8
+    {
+        const int intResult = static_cast<int>(lhs) - static_cast<int>(rhs);
+        return static_cast<Uint8>(intResult > 0 ? intResult : 0);
+    };
+
+    return Color(clampedSub(left.r, right.r),
+                 clampedSub(left.g, right.g),
+                 clampedSub(left.b, right.b),
+                 clampedSub(left.a, right.a));
+}
+
+
+////////////////////////////////////////////////////////////
+constexpr inline Color operator *(const Color& left, const Color& right)
+{
+    const auto scaledMul = [](Uint8 lhs, Uint8 rhs) -> Uint8
+    {
+        const int intResult = static_cast<int>(lhs) * static_cast<int>(rhs);
+        return static_cast<Uint8>(intResult / 255);
```

Comment:
On very weird architectures where `int` has 16 bits, this could cause an overflow: `255 * 255 > 2^(16-1)`

I know this was already a problem before, so not sure if it would change anything in practice. In any case, `Uint16` should do the trick.

Reply:
Honestly curious -- is there any such platform on which SFML could be used in this day and age? I will make the change anyway, it seems like a good idea to be explicit with our signedness and boundaries here.

Reply:
I've applied the change, but I'm not sure if it's an improvement compared to what we had before. We should consider reverting unless there actually is a platform where `int` has 16 bits which is widely used by SFML

Reply:
I'm not aware of such platforms. Likely, this would also have led to problems elsewhere already...

In C++, there's on one side the standard crusaders who made correcting each other on StackOverflow their hobby, and on the other the "it works" people relying on various UB and implementation-defined semantics (gamedev is a great example). And then a range in between 😬

But since so many things are loosely specified, it's sometimes really hard to know whether something is a real issue in practice or not.

> We should consider reverting unless there actually is a platform where `int` has 16 bits which is targeted by SFML.

Do we need to be as overly verbose at all? Otherwise, let's use integral promotion (`Uint8 * Uint8 == int`).
```cpp
int result = lhs * rhs;
return static_cast<Uint8>(result / 255);
```
Or does that issue a warning?

Reply:
We will get warnings without casts, but using `int` instead of `Uint16` does require less casts, assuming that we do not care about platforms using 16-bit `int`. How would you like me to proceed? IMHO, we should not support such platforms unless we get explicitly requested to do so.

Reply:
I tend to agree, however that would raise the question why we have `sf::Uint8` etc. in the first place (or `uint8_t`, that's not the point).

There are two main approaches:
1. We embrace the fact that standard types are variable-sized and don't rely on "99%" assumptions. We use fixed-size types where we need them. This might be the safest and most portable option, plus it could be more expressive about the underlying assumptions in code.
2. We deliberately restrict SFML to certain platforms (possibly enforced via preprocessor, e.g. `#error` on wrong `INT_MAX` value or so). We should only do this after deeper research however, as it's really not nice to restrict certain valid use cases for "cosmetic" reasons. If we go for this, we should also be pragmatic w.r.t. other "common" implementations (`ptrdiff_t` etc).

I think it's a bigger topic, so maybe let's discuss it elsewhere.

About here, I don't know. Reducing/beautifying casts wasn't exactly a concern during https://github.com/SFML/SFML/pull/1846, so maybe we should still go for the more verbose way with `Uint16` here?

Reply:
> maybe we should still go for the more verbose way with Uint16 here?

I am personally fine with that.